### PR TITLE
Bugfix/lazy component loading states and fetching

### PR DESCRIPTION
**Pull request by filrak**
### Short description and why it's useful

- added ability to reload component when fetch fails
- added loading states for async sidebar components
- added error handling for offline/online errors

TODO
- [x] add left sidebar support
- [ ] add custom width support
- [x] refactor transitions
- [x] add `close` button on failed load
- [x] preload crucial async components
- [ ] add translation keys
### Screenshots of visual changes before/after (if there are any)
(Failed loading - is offline)
image
(Failed loading - is online)
image
Component is loading
image

The solution is still dirty and needs polishment. Just leaving the WIP PR to mostly show screenshots :)

### CODE REVIEW

**Comment by pkarw** [comment]
This is pretty important fix and should be applied even when not polished. However - the Micorcart, Menu + Wishlist sidebars should be eagerly loaded for the offline mode - this is something i discussed with @patzick and there is a ticket for that. So I'd preffer to have them eagerly loaded this week and then think off what's left for this fancy loader You created here @filrak - we should have both!

**Comment by filrak** [comment]
It will be finished today. With this approach user will be able to refresh failed component just with one click (we can even try to do this automatically after connection is restored).

I'm not that a big fan of eager loading of this components. No matter if we have eager load or prefetch they still need to be downloaded anyway and in uncertain network conditions the bigger initial bundle is the worse for us.

What is the exact reason of eager load of this offscreen components?

@patzick ?

**Comment by pkarw** [comment]
Go to product page, swithc internet off, add product to the cart -try checking the cart status :) Good luckl;  at least for now ;)

That's the reason for eagerly loading the key components - just cart, wishlist + navigation.

**Comment by filrak** [comment]
but this PR will solve this problem

**Comment by filrak** [comment]
Ok, then this PR is prty useless in core functionalities. More like an additional feature

**Comment by pkarw** [comment]
We should have it + Microcart + navigation eagerly loaded. I really like this feature - let's merge it in

**Comment by filrak** [comment]
I added automatic reload of failed components when user cames back online

**Comment by pkarw** [comment]
Ok; fine with me

**Comment by filrak** [comment]
Done. I will improve some parts during stability phase but for now it works perfectly fine and solves the problem gracefully ;).

File: src/themes/default/components/theme/blocks/AsyncSidebar/AsyncSidebar.vue [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+// todo: add left sidebar option
+export default {
+  props: {
+    asyncComponent: {
+      type: Function,
+      required: true
+    },
+    isOpen: {
+      type: Boolean,
+      required: true
```
Comment (patzick) [inline src/themes/default/components/theme/blocks/AsyncSidebar/AsyncSidebar.vue]:
```suggestion
      default: false
```

File: src/themes/default/components/theme/blocks/AsyncSidebar/AsyncSidebar.vue [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    },
+    /** Action that closes the modal. should set value referenced by `isOpen` to false */
+    close: {
```
Comment (patzick) [inline src/themes/default/components/theme/blocks/AsyncSidebar/AsyncSidebar.vue]: this props should be removed - replaced above

File: src/themes/default/components/theme/blocks/AsyncSidebar/AsyncSidebar.vue [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      type: Function,
+      required: true
+    },
+    /** "right" or "left"  */
+    direction: {
+      type: String,
+      required: true
```
Comment (patzick) [inline src/themes/default/components/theme/blocks/AsyncSidebar/AsyncSidebar.vue]:
```suggestion
      default: 'left'
```

File: src/themes/default/components/theme/blocks/AsyncSidebar/AsyncSidebar.vue [hunk 4]
```diff
@@ -0,0 +1,116 @@
+<template>
+  <transition :name="direction === 'right' ? 'slide-left' : direction === 'left' ? 'slide-right' : null ">
+    <div
+      class="mw-100 fixed cl-accent bg-cl-primary"
+      :class="direction === 'left' ? 'left-sidebar' : direction === 'right' ? 'right-sidebar' : null "
+      data-testid="sidebar"
+      v-if="isOpen"
+    >
+      <component :is="component"/>
```
Comment (patzick) [inline src/themes/default/components/theme/blocks/AsyncSidebar/AsyncSidebar.vue]:
```suggestion
      <component :is="component" @close="$emit('close')" @reload="getComponent" />
```

File: src/themes/default/components/theme/blocks/AsyncSidebar/LoadingError.vue [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  name: 'LoadingError',
+  mixins: [VueOfflineMixin],
+  mounted () {
+    this.$on('online', () => {
+      this.reload()
+    })
+  },
+  methods: {
+    reload () {
+      this.$parent.getComponent()
```
Comment (patzick) [inline src/themes/default/components/theme/blocks/AsyncSidebar/LoadingError.vue]:
```suggestion
      this.$emit('reload')
```

File: src/themes/default/components/theme/blocks/AsyncSidebar/LoadingError.vue [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    },
+    close () {
+      this.$parent.close()
```
Comment (patzick) [inline src/themes/default/components/theme/blocks/AsyncSidebar/LoadingError.vue]:
```suggestion
      this.$emit('close')
```

File: src/themes/default/layouts/Default.vue [hunk 7]
```diff
@@ -4,14 +4,30 @@
...
-      </transition>
-      <transition name="slide-left">
-        <microcart v-if="isMicrocartOpen"/>
-        <search-panel v-if="isSearchPanelOpen"/>
-        <wishlist v-if="isWishlistOpen"/>
-      </transition>
+      <async-sidebar
+        :async-component="SearchPanel"
+        :is-open="isSearchPanelOpen"
+        :close="() => $store.commit('ui/setSearchpanel', false)"
```
Comment (patzick) [inline src/themes/default/layouts/Default.vue]:
```suggestion
        @close="$store.commit('ui/setSearchpanel')"
```

File: src/themes/default/layouts/Default.vue [hunk 8]
```diff
@@ -4,14 +4,30 @@
...
+        direction="right"
+      />
+      <async-sidebar
+        :async-component="Microcart"
+        :is-open="isMicrocartOpen"
+        :close="() => $store.commit('ui/setMicrocart', false)"
```
Comment (patzick) [inline src/themes/default/layouts/Default.vue]:
```suggestion
        @close="$store.commit('ui/setMicrocart')"
```

File: src/themes/default/layouts/Default.vue [hunk 9]
```diff
@@ -4,14 +4,30 @@
...
+        direction="right"
+      />
+      <async-sidebar
+        :async-component="SidebarMenu"
+        :is-open="isSidebarOpen"
+        :close="() => $store.commit('ui/setSidebar', false)"
```
Comment (patzick) [inline src/themes/default/layouts/Default.vue]:
```suggestion
        @close="$store.commit('ui/setSidebar')"
```

File: src/themes/default/layouts/Default.vue [hunk 10]
```diff
@@ -4,14 +4,30 @@
...
+        direction="left"
+      />
+      <async-sidebar
+        :async-component="Wishlist"
+        :is-open="isWishlistOpen"
+        :close="() => $store.commit('ui/setWishlist', false)"
```
Comment (patzick) [inline src/themes/default/layouts/Default.vue]:
```suggestion
        @close="$store.commit('ui/setWishlist')"
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -30,6 +30,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - add cart count config, allows you to display the item count instead of a sum of the item quantities - @pauluse (#2483)
 - improved product gallery load view, shows correct image on reload - @patzick (#2481, #2482, #2488, #2501)
 - Fix an issue where the index.html template within a theme is ignored - @EnthrallRecords (#2489) 
+- Added async sidebar component with async off-screen components error handling and fetch retrying after coming back online - @filrak (#2500)
```
Comment (patzick) [inline CHANGELOG.md]:
```suggestion
- Added async sidebar component with async off-screen components error handling and fetch retrying after coming back online - @filrak (#2408, #2451)
```

**Comment by patzick** [comment]
CR does not link issues so i need a separate comment - related issues #2408, #2451

**Pull request merged**